### ISSUE TITLE
cfg(test) is not set during integration tests

### ISSUE THREAD

**Issue by Timmmm** [rust-lang/rust#84629]
`#[cfg(test)]` is not set when running [integration tests](https://doc.rust-lang.org/book/ch11-03-test-organization.html#integration-tests). That page says:

> [Integration tests] use your library in the same way any other code would, which means they can only call functions that are part of your library’s public API.

But that sounds to me more like it's talking about `pub`, `pub(crate)` and so on.

There's [an unsatisfying workaround](https://stackoverflow.com/a/47698997/265521) but it would be a lot simpler if `#[cfg(test)]` was simply enabled for integration tests.

The only reason not to do that seems to be a philosophical objection to using hidden APIs in integration tests. I don't agree with that. I'm sure others don't. But even if you subscribe to that philosophy you can simply not use the `#[cfg(test)]` functions in your integration tests (or don't export them). The situation for people who don't subscribe to that philosophy is much more awkward.

Changing this behaviour would also fix #45599.

**Comment by hyd-dev** [comment]
> `#[cfg(test)]` is not set when running [integration tests](https://doc.rust-lang.org/book/ch11-03-test-organization.html#integration-tests).

I believe `#[cfg(test)]` means "the test harness is used" (i.e. `#[test]` is active): https://doc.rust-lang.org/nightly/reference/conditional-compilation.html#test. Setting it in a library where `#[test]` is not active would be very unexpected to me. I'd suggest a different `cfg` like `#[cfg(integration_test)]`.

**Comment by Timmmm** [comment]
That's a great idea, than you can do `any(cfg(test), cfg(integration_test))` if you want.

**Comment by Timmmm** [comment]
Although a significant downside is that it adds another variant of the crate that you have to compile, which uses more disk space & time and makes things more complicated for IDEs.

**Comment by hyd-dev** [comment]
> Although a significant downside is that it adds another variant of the crate that you have to compile, which uses more disk space & time and makes things more complicated for IDEs.

That downside it not specific to `integration_test` though. You will have two variants even if you use `test`:
- a binary crate variant that runs all `#[test]`s
- a library crate variant with `#[cfg(test)]` enabled but `#[test]` disabled

**Comment by Timmmm** [comment]
Good point. Perhaps the standard `#[test]` compilation could compile it as a library too (and then link with the test runner).

**Comment by epage** [comment]
This sounds like its about setting `--cfg test` on all crates a test binary depends on (including the lib within the same package).  If so, this is less about `libtest` and more about `cargo test` and would belong in the cargo repo.  In fact, we have rust-lang/cargo#8379 open for this.
